File: leetcode_scala/src/com/leetcode/book/Offer_05_替换空格.scala
package com.leetcode.book

/**
 * Author:   Yang
 * Date:     2021/5/28 
 * Description:
 * 请实现一个函数，把字符串 s 中的每个空格替换成"%20"。
 * 示例 1：
 * 输入：s = "We are happy."
 * 输出："We%20are%20happy."
 */
object Offer_05_替换空格 {
  def main(args: Array[String]): Unit = {
    val s = "We are happy."
    val res: String = replaceSpace(s)
    println(res)

  }

  def replaceSpace(s: String): String = {

    // 第一种方法 split => reduce    (有缺陷：当只有一个元素时，reduce会发出异常)
    val str: Array[String] = s.split("\\s+")
    str.reduce((x1, x2) => (x1 + "%20" + x2))


  }


}

File: leetcode_java/src/com/leetcode/book/Offer_05_替换空格.java
/**
 * Copyright (C), 2015-2021, XXX有限公司
 * FileName: Offer_05_替换空格
 * Author:   Hankesstill
 * Date:     2021/5/28 10:56
 * History:
 * <author>          <time>          <version>          <desc>
 * 作者姓名           修改时间           版本号              描述
 */
package com.leetcode.book;

/**
 * @author Hankesstill
 * @date 2021/5/28 10:56
 * 请实现一个函数，把字符串 s 中的每个空格替换成"%20"。
 * 示例 1：
 * 输入：s = "We are happy."
 * 输出："We%20are%20happy."
 */
public class Offer_05_替换空格 {
	public static void main(String[] args) {
		String s = new String("We are happy");
		String res = replaceSpace(s);
		System.out.println(res);
	}

	public static String replaceSpace(String s) {

		/**
		 * 1  遍历添加
		 * 		创建一个 StringBuffer
		 * 		依次向内部添加元素
		 *
		 */
		StringBuffer st = new StringBuffer();
		for (char c : s.toCharArray()) {
			if (' ' == c) {
				st.append("%20");
			} else {
				st.append(c);
			}
		}

		return st.toString();

	}

}